### ISSUE TITLE
ttnn.rms_norm on fp32 inputs is less accurate than an equivalent manual reduction

### ISSUE THREAD

**Issue by iwroszTT** [tenstorrent/tt-metal#49600]
## Summary

For fp32 inputs, the ttnn.rms_norm is roughly 1–2 orders of magnitude less accurate than computing the identical RMS/L2 normalization by hand with a multiply, reduce, rsqrt, multiply sequence. 

For input [1, 2048, 128], FLOAT32, TILE, normalizing over the last dim:
```

  ┌────────────────────────────────────────────────────────────────────────┬─────┬───────────────┐
  │                                 method                                 │ ops │ max abs error │
  ├────────────────────────────────────────────────────────────────────────┼─────┼───────────────┤
  │ multiply, mean/sum, add, rsqrt, multiply                               │   5 │       ~8.6e-5 │
  ├────────────────────────────────────────────────────────────────────────┼─────┼───────────────┤
  │ ttnn.rms_norm, default compute config                                  │   1 │       ~5.5e-3 │
  ├────────────────────────────────────────────────────────────────────────┼─────┼───────────────┤
  │ ttnn.rms_norm — HiFi4 + fp32_dest_acc_en=True + math_approx_mode=False │   1 │      ~1.07e-3 │
  └────────────────────────────────────────────────────────────────────────┴─────┴───────────────┘
```

ttnn.rms_norm should be within a small factor of a manual fp32 reduction.

Found while working on the linear attention block in Qwen3.5
https://github.com/tenstorrent/tt-metal/blob/b9547f9a81c24797a16982430af8f2fbe4969788/models/experimental/gated_attention_gated_deltanet/torch_functional/delta_rule_ops.py#L26

## Test case:
```

import pytest
  import torch
  import ttnn

  pytestmark = pytest.mark.use_module_device


  def _torch_rms(x, eps):
      # float64 reference: x / sqrt(mean(x^2) + eps)
      xd = x.double()
      return (xd / torch.sqrt(xd.pow(2).mean(dim=-1, keepdim=True) + eps)).float()
  
  
  def _manual_rms_ttnn(x, eps, cfg):
      # Hand-rolled RMS: multiply -> mean -> add -> rsqrt -> multiply (the accurate baseline).
      x_sq = ttnn.multiply(x, x)
      ms = ttnn.mean(x_sq, dim=-1, keepdim=True)
      inv = ttnn.rsqrt(ttnn.add(ms, eps))
      return ttnn.multiply(x, inv)

  
  @pytest.mark.parametrize("h", [2048])
  @pytest.mark.parametrize("w", [128, 4096])
  @pytest.mark.parametrize("dtype", [torch.float32, torch.bfloat16])
  def test_rms_norm_fp32_accuracy_vs_manual(device, h, w, dtype):
      torch.manual_seed(0)
      eps = 1e-6

      torch_x = (torch.randn(1, h, w) * 1.7).to(dtype)
      golden = _torch_rms(torch_x, eps)
  
      # HiFi4 + fp32 dest acc: best-case accuracy for the fused op.
          device.arch(),
          math_fidelity=ttnn.MathFidelity.HiFi4,
          math_approx_mode=False,
          fp32_dest_acc_en=True,
      )
      
      x = ttnn.from_torch(torch_x, device=device, layout=ttnn.TILE_LAYOUT)
      ones = ttnn.from_torch(torch.ones(w, dtype=dtype), device=device, layout=ttnn.TILE_LAYOUT)

      manual = ttnn.to_torch(_manual_rms_ttnn(x, eps, cfg)).float()
      fused = ttnn.to_torch(
          ttnn.rms_norm(x, epsilon=eps, weight=ones, compute_kernel_config=cfg)
      ).float()
      
      manual_err = (manual - golden).abs().max().item()
      fused_err = (fused - golden).abs().max().item()
      print(f"[{dtype}] w={w}  manual_max_err={manual_err:.3e}  fused_max_err={fused_err:.3e}
  ratio={fused_err/max(manual_err,1e-12):.1f}x")
  
      # Expectation: the fused op should be within ~3x of the manual reduction.
      # On FLOAT32 this currently FAILS (fused ~10-60x worse) -> the bug.
      assert fused_err <= 3.0 * manual_err, (
          f"ttnn.rms_norm ({dtype}) max abs err {fused_err:.3e} is "
          f"{fused_err/manual_err:.1f}x the manual reduction's {manual_err:.3e}"
      )
```

**Comment by FortMyersBrewing** [comment]
I'm going to preface with - I'm BRAND new at this codebase, my p150's are literally in transit, but I've been hard at it with sim and coding and I've dug into this on ttsim (Blackhole). Short version: `rms_norm` misses two of the accuracy measures laid out in `tech_reports/op_kernel_dev/accuracy_tips/accuracy_tips.md`, which `layer_norm` and `group_norm` do apply. One is a one-line default worth ~5x. The other is structural and accounts for the rest.

(Heads up: the test as posted doesn't run - the `cfg = ttnn.<...>(` line is missing from the code block, leaving a dangling arg list. I used `ttnn.init_device_compute_kernel_config(device.arch(), ...)`, which fits your arguments, let me know if I used something wrong here)

| method | yours | ratio | mine (BH ttsim) | ratio |
|---|---|---|---|---|
| manual (mul, mean, add, rsqrt, mul) | 8.6e-5 | 1x | 8.03e-4 | 1x |
| ttnn.rms_norm, default compute config | 5.5e-3 | 64x | 4.04e-2 | 50x |
| ttnn.rms_norm, HiFi4 + fp32_dest_acc_en | 1.07e-3 | 12.4x | 8.81e-3 | 11.0x |

Same shape as yours. w=4096 is worse still (16.4x), and bf16 passes at both widths - fused is actually better than manual there, which is the tell that this is about fp32 not fitting somewhere.

### 1. rms_norm defaults fp32_dest_acc_en to false; layer_norm defaults it to true

```cpp
// rmsnorm.cpp:16                    // layernorm.cpp:15
bool approx_mode = true;             bool approx_mode = false;
bool fp32_acc    = false;            bool fp32_acc    = true;
```

accuracy_tips.md:56-58 is explicit that `fp32_dest_acc_en` is what puts DEST in fp32, and :79 that failing to configure fp32 accumulation "can lead to large, noisy errors" - with LayerNorm ULP plots showing exactly that. With fp32_acc off, every intermediate CB also drops to bfloat16 (`layernorm_op_multi_core.cpp:151`).

The divergence is plain in the source regardless of where you run it. Confirming in sim: calling rms_norm with no config is bit-identical to calling it with `rmsnorm_default_compute_config` as written, and swapping in layernorm's default takes 4.04e-2 to 8.81e-3 - a 4.6x win. Your table shows the same jump (5.5e-3 to 1.07e-3). Given the sibling op does the opposite, this looks unintentional, and it's fixable on its own.

### 2. The per-element math runs through srcA/B, which can't carry fp32

This is the part that survives fixing the default. From `base_types.hpp:31-39`:

> Default mode enables all dataformats (except Float32) to be unpacked into Dest. [...] UnpackToDestFp32 enables unpacking Float32 data to Dest with full precision, but makes the buffer incompatible with unpacking to SRCA/B.

and accuracy_tips.md:58: "the data will be converted to TF32 before being copied to dest, gradually eroding the accuracy."

In sim, computing `y = x` two ways on identical fp32 data:

```
SFPU  x*1  (ttnn.multiply): bit-exact,        0/262144 elements changed
FPU   x@I  (ttnn.matmul)  : lossy,       262111/262144 changed
```

Your 5-op manual chain stays accurate because `ttnn.multiply`/`add`/`rsqrt` route fp32 to the SFPU with UnpackToDestFp32 (`binary_ng_program_factory.cpp:745`, `unary_program_factory.cpp:171`). Both paths still use the FPU for the reduce, but there the error is one term in a sum over W and averages down by ~sqrt(W). rms_norm's per-element multiply (`layernorm.cpp:283`, `mul_tiles_bcast_cols`) doesn't average, it lands on every output! That asymmetry is the 11x.

Note the reduce itself is fine - `row_wise_mean` already does the sum-then-divide from accuracy_tips.md:89-112 (scalar tile of 1s, then `scale_dest(1/N)`). It's specifically the per-element multiply.

Worth flagging: accuracy_tips.md:77 says a CB with UnpackToDestFp32 "can not be used in operations that use the srcA/B registers", so this is not a matter of flipping the flag on the input CB. The per-element ops have to move to `copy_tile` + SFPU binary in DEST, the way `binary_ng` does. Not free - SFPU eltwise is slower than `mul_tiles` so I'm guessing it likely warrants gating on fp32 inputs.

Also, rms_norm can't currently use Welford (accuracy_tips.md:141), which layer_norm and group_norm can. Not the cause here, but it's the third item on that checklist it doesn't get.

One more: passing `weight=ones` roughly doubles the error (8.81e-3 vs 4.76e-3 with weight=None), because it adds a gamma multiply - another srcA/B pass over the already normalized result. Your repro passes ones, so about half of what's left is that one redundant multiply.

### Caveat, and a request

All my absolutes run 7-9x higher than yours. Backing the implied per-element precision out of the fused numbers (max |output| is 4.46):

```
yours: 1.07e-3 / 4.46 = 2.4e-4  ->  ~12 mantissa bits
mine:  8.81e-3 / 4.46 = 2.0e-3  ->  ~9  mantissa bits
```

A consistent ~3-bit (8x) offset, which would explain all three rows at once. Measuring the FPU's effective operand width on ttsim I get ~10 bits at HiFi4, and on that number your 1.07e-3 is better than an srcA/B multiply should be able to produce. So either ttsim's FPU is a few bits coarser than silicon, or your build's rms_norm isn't taking the path I think it is. The qualitative story is unaffected either way (~12 bits is still far from fp32's 24, and is exactly what a narrow srcA/B operand looks like), but I don't want to assert bit widths from a simulator.

If someone with a card can run an fp32 `matmul(x, I)` and count how many mantissa bits survive, that settles it. And could you share your commit and arch? That would pin down which of the two it is.

**Comment by bbradelTT** [comment]
@fplavecTT please take a look since you're dealing with https://github.com/tenstorrent/tt-metal/issues/48824

**Comment by FortMyersBrewing** [comment]
Follow-up on my own comment (yes I pretty much stayed up all night down this long rabbit hole!) one thing I got too cautious about, one thing I need to walk back, and a question. Still very much learning this codebase, so push back where I've got it wrong.

**The ~10 mantissa bits isn't a simulator artifact, I was too cautious.** I ran an fp32 rms_norm and read the `chlkc_descriptors.h` that the JIT emits into the kernel cache. With `fp32_dest_acc_en=True`, verbatim:

```c
constexpr uint8_t unpack_src_format[64] = { 0,255,5,5,255,0, ... };
constexpr uint8_t unpack_dst_format[64] = { 4,255,5,5,255,4, ... };
```

Index 0 is cb0, the input CB (`layernorm_op_multi_core.cpp:681`). `0` is Float32 and `4` is Tf32 (`tt_backend_api_types.hpp:47`). And `unpack_dst_format` is the format the unpacker converts an operand into on its way to SrcA - `llk_unpack_common.h:132` ("New destination data format operand A is converted to"), written to `THCON_SEC0_REG2_Out_data_format` at `:180`.

So the fp32 input becomes Tf32 in SrcA. Tf32 is 10 mantissa bits, which matches what I measured black-box.

Worth being clear about why this isn't just a ttsim quirk: that header is generated host-side by the JIT, and the selection in `genfiles.cpp:624-629` branches only on the buffer formats and `fp32_dest_acc_en` - nothing about the target. So the same constant lands in the kernel whether you compile for the simulator or for a card. I've spent way too many hours trying to figure out if this was a sim compile thing (see walking back the bf16 line at the bottom.)

**And with the default config it's worse than I said.** Same op, same file, but with rms_norm's actual default (`fp32_dest_acc_en=False`):

```c
constexpr uint8_t unpack_src_format[64] = { 0,255,5,5,255,0, ... };
constexpr uint8_t unpack_dst_format[64] = { 5,255,5,5,255,5, ... };
```

cb0 now goes `0 -> 5`: Float32 to **Float16_b**. So by default `ttnn.rms_norm` converts an fp32 input to bfloat16 - 7 mantissa bits on its way into SrcA, before any math runs. Same `unpack_conditional_dst_format` selection as above, just the other branch. Which means sections 1 and 2 of my earlier comment aren't two problems: it's one mechanism (fp32 can't live in SrcA, so it gets converted), and `fp32_dest_acc_en` only decides whether that conversion is Tf32 (10 bits) or bf16 (7 bits).

The default: `fp32_dest_acc_en=False` seems a reasonable performance choice for bf16 workloads, which are presumably the common case. The issue doesn't seem to be is the default right or wrong - it's that for a **Float32 input** it quietly drops you to bf16 operands, and nothing tells you.

**A question I might just be missing context on.** Both group_norm and layer_norm guard against `Float32 && !fp32_dest_acc_en`, though each adds a further condition:

```cpp
// groupnorm_mcast_program_factory.cpp:236  (also requires use_welford)
"group_norm welford with Float32 input requires fp32_dest_acc_en=true ...
 otherwise precision is silently lost in the unpacker format conversion."

// layernorm_op_multi_core.cpp:331  (also requires input_is_row_major)
"ROW_MAJOR float32 input requires fp32_dest_acc_en=True ...
 (SRCA is 16-bit when fp32_dest_acc_en=False, silently truncating float32 to bfloat16 ...)"
```

Neither reaches rms_norm - the first is gated on `use_welford` (which rms_norm doesn't use), the second on `input_is_row_major` (this repro is TILE). Is there a reason rms_norm doesn't have an equivalent, or is it just that nobody's needed one yet? One data point that might be relevant: the layer_norm message attributes the truncation to tilize, but the descriptor dump above shows it also happens for TILE inputs, so that guard may cover less than it looks like it does.

If it's worth doing something here, making the default dtype-aware (fp32 input -> fp32_dest_acc_en=True) seems less disruptive than flipping it outright, since bf16 users presumably want the current behaviour.

**Walking back the bf16 line.** I said bf16 passes at both widths. That's true in ttsim, but I've since found ttsim does not model the packer's round-vs-truncate control - forcing `PCK_DEST_RD_CTRL_Round_10b_mant` gives `UnimplementedFunctionality: tensix_pacr`, and with it off ttsim rounds where hardware truncates (which is the premise of #49111). bf16 output packs from an fp32 dest down to bf16, exactly that path, so please don't lean on my bf16 numbers. The fp32 rows are unaffected - with `fp32_dest_acc_en=True` the intermediates are Float32, so those packs don't narrow and the question never arises.

**On the ~8x gap between our absolutes.** Since the Float32 -> Tf32 conversion is decided host-side, your hardware feeds SrcA a 10-mantissa-bit operand too, which means your fused number (implying ~12 bits) probably isn't coming from an srcA/B multiply at all. Of the two explanations I floated, "ttsim's FPU is coarser than silicon" now looks unlikely, and "the rms_norm I'm looking at isn't the path you're hitting" looks more likely. Happy to re-run anything against a different revision if that's useful.

Cross-referencing #49111 - @friedrich found layer_norm's intermediate truncation errors were accidentally cancelling, which looks like another face of the same "fp32 doesn't survive the pipeline" problem.

**Comment by friedrich** [comment]
It turns out that the cancelling was just a coincidence that pushed SDPA into a regime where it was more accurate. The underlying issue was using bfloat16 for the scores in SDPA: https://github.com/tenstorrent/tt-metal/pull/49948

**Comment by AJ0070** [comment]
Independent reproduction from a different model family, plus two data points that may help with prioritisation: how the gap scales with the normalised dimension, and what it costs a real model end to end.

I hit this while bringing up Granite 4.0-H (hybrid Mamba-2) on an N300, chasing an accuracy loss at depth. Same conclusion as the report: the manual multiply / reduce / rsqrt / multiply sequence is far more accurate than `ttnn.rms_norm` on fp32.

**The gap widens with the normalised dimension.** The report tests last-dim 128; LLM hidden sizes are 2048 and up. Max **relative** error against a float64 oracle (note the report quotes absolute error, so these are not directly comparable to its table):

| normalised dim | input | `ttnn.rms_norm` | manual reduction |
| --- | --- | --- | --- |
| 2048 | `randn` | 3.587e-02 | 1.798e-07 |
| 4096 | `randn` | 4.939e-02 | 1.923e-07 |
| 2048 | with outlier channels | 6.615e-02 | 2.799e-07 |
| 4096 | with outlier channels | 1.155e-01 | 2.961e-07 |

At 128 the report measures roughly 1 to 2 orders of magnitude. At 2048 to 4096 it is 5 to 6, and it grows monotonically with the dimension.

The "outlier channels" rows scale every 512th channel by 40x, which is the shape trained residual streams actually have rather than clean `randn`. That roughly doubles the error again, so the synthetic case is the optimistic one.

**End to end the effect is much smaller, and I think that matters for priority.** Substituting `ttnn.rms_norm` back into the full 40-layer model on real `granite-4.0-h-micro` weights:

| | logits PCC vs reference | top-1 agreement |
| --- | --- | --- |
| manual reduction | 0.999310 | 100.0% |
| `ttnn.rms_norm` | 0.997732 | 100.0% |

The model still predicts the same token. A 40-layer stack applies a norm about 100 times, so the per-application error does compound, but activations are bfloat16, which keeps the fp32 path from being fully exercised. So: a large op-level defect with a modest model-level consequence for a bf16 inference stack, and presumably a much larger one for anything genuinely running fp32 activations, which is where the original Qwen3.5 report came from.

Happy to share the repro script or run further shapes on the N300 if that would help.
